#106 Added filtering and cache to manage glosses

TheTuTran opened the pull request:
I had issues setting up the Dockerfile, but there was a problem with my chrome caching which was the cause of my issues. 
I attempted to add a cache and filter to copy the code from glossing.html, but it obviously didn't work. Going to try to re-read deer-config to see what's going on.

edit: Caching and Filtering should work. After a user clicks the visibilitybtn and "Save List" button, the list should update the public list on the server. 
Here's the dev link https://devstore.rerum.io/v1/id/610c54deffce846a83e70625 

Before: 
![image](https://github.com/CenterForDigitalHumanities/glossing-entries/assets/131922588/7051e7c6-aae2-4d7e-9f7b-13c5661f8402)

After:
![image](https://github.com/CenterForDigitalHumanities/glossing-entries/assets/131922588/a3f08bf4-8f30-4388-adf6-6dbf9a896268)
![image](https://github.com/CenterForDigitalHumanities/glossing-entries/assets/131922588/c534cc2e-706e-417d-8928-aa2bf1f0a98a)

AbhilashKotha reviewed (changes requested):
@TheTuTran The functionality works. I just have the following feedback

1. Save list does not give the user any message - success or failure. 
2. This happens for me on initial load. what do you think is the issue?
3. I see that you committed Docker file do we need that? what is Prettier ignore file?
<img width="1280" alt="image" src="https://github.com/CenterForDigitalHumanities/glossing-entries/assets/113061137/7984fb2b-3563-4e2e-8718-14a019926141">

TheTuTran commented:
> @TheTuTran The functionality works. I just have the following feedback
> 
> 1. Save list does not give the user any message - success or failure.
> 2. This happens for me on initial load. what do you think is the issue?
> 3. I see that you committed Docker file do we need that? what is Prettier ignore file?
> 
> <img alt="image" width="1280" src="https://user-images.githubusercontent.com/113061137/284163434-7984fb2b-3563-4e2e-8718-14a019926141.png">

1. I added a prompt for successful save and unsuccessful save.
2. I don't know what is causing that. I've just been assuming it was some sort of browser issue because it happens to me unexplainably. I think it's something with the gloss.css since the checkmark is created from ".is-included". 
3. I removed the dockerfile, and the Prettierignore is for my vscode extension (My extension would automatically break the functions up into separate lines & added semicolons at the end of every line).

thehabes commented:
2. is happening because your browser is being a jerk and isn't rendering UTF-8.  If you use something like https://chromewebstore.google.com/detail/charset/oenllhgkiiljibhfagbfogdbchhdchml and tell the page to be UTF-8 this will stop happening.

thehabes reviewed (changes requested):
**You have caused duplicates to reappear.**

**glosses.html**
![image](https://github.com/CenterForDigitalHumanities/glossing-entries/assets/3287006/344870ef-b953-4b3f-9374-8dabbbf3bbfc)

**gloss-transcription.html?tpen-project=https://t-pen.org/TPEN/project/4080**
![image](https://github.com/CenterForDigitalHumanities/glossing-entries/assets/3287006/16244e1f-c18d-4b82-b111-dcb677fc4079)

**Duplicates appear in your list**

**manage-glosses.html**
![image](https://github.com/CenterForDigitalHumanities/glossing-entries/assets/3287006/c1919bc2-6c46-400a-b454-e80dd182ee32)

Make sure you merge main into your branch and take care of conflicts.  There are deployed bug fixes and features upstream from you that your code needs.

AbhilashKotha commented:
@TheTuTran, I suggest you use the global feedback blip for the messages. @thehabes what do you say?

something like this
<img width="683" alt="image" src="https://github.com/CenterForDigitalHumanities/glossing-entries/assets/113061137/2c56d9de-b3cc-46b6-833b-e2661ce778b2">
<img width="1274" alt="image" src="https://github.com/CenterForDigitalHumanities/glossing-entries/assets/113061137/a00eb218-026f-4a66-a0ce-269d4d9a1e00">

thehabes commented on js/deer-config.js:
```diff
@@ -648,13 +650,287 @@ export default {
+            return {
+                html: ``,
+                then: (elem) => {
+                    let cachedFilterableEntities = localStorage.getItem("managedListCache") ? new Map(Object.entries(JSON.parse(localStorage.getItem("managedListCache")))) : new Map()
```
You need to use the same localStorage entry as the other templates, otherwise they won't share the cache.  All the gloss list related templates can use the same cache.

thehabes commented on js/deer-render.js:
```diff
@@ -1043,70 +1043,249 @@ DEER.TEMPLATES.managedlist = function (obj, options = {}) {
+            <p class="filterNotice is-hidden"> Gloss filter detected.  Please note that Glosses will appear as they are fully loaded. </p>
+            <div class="totalsProgress" count="0"> {loaded} out of {total} loaded (0%).  This may take a few minutes.  You may click to select any Gloss loaded already.</div>
+        </div>`
+        let managedListCache = localStorage.getItem("managedListCache") ? new Map(Object.entries(JSON.parse(localStorage.getItem("managedListCache")))) : new Map()
```
You need to use the same localStorage entry as the other templates, otherwise they won't share the cache. All the gloss list related templates can use the same cache.

thehabes reviewed (changes requested):
There are still some issues here. 

1.  Starting with an empty cache, after letting the list cache on managed-glosses.html, it should also be cached for ng.html and gloss-transcription.html.  When I tested last time I thought it was, but it is not.  After letting it cache, when I go to one of these pages, I have to wait for it to cache again and vice-versa.
1. Duplicates still appear in the list on manage-glosses.html

Everything else is cleaned up, but you need to get this functionality right before we can merge.

thehabes commented on js/deer-render.js:
```diff
@@ -1043,70 +1043,249 @@ DEER.TEMPLATES.managedlist = function (obj, options = {}) {
+        let numloaded = 0
         const type = obj.name.includes("Named-Glosses") ? "named-gloss" : "manuscript"
 
+        const total = obj[options.list].length
```
Duplicates still happening.

thehabes commented on js/deer-config.js:
```diff
@@ -373,22 +374,24 @@ export default {
             let numloaded = 0
             let total = 0
-            const filterPresent = deerUtils.getURLParameter("gog-filter")
+            const filterPresent = !!deerUtils.getURLParameter("gog-filter")
```
oo have these been coming from your linter!?!?!  Get that `!!` outta here

thehabes replied:
See deer-config.js for more instances of `!!`

thehabes commented on js/deer-config.js:
```diff
@@ -373,22 +374,24 @@ export default {
                     let inclusionBtn = null
                     const glossID = val['@id'].replace(/^https?:/, 'https:')
                     let already = witnessesObj?.referencedGlosses?.has(glossID) ? "attached-to-source" : ""
+                    console.log("glossID from glossesSelectorForTextualWitness", glossID)
```
Clear out the log.

thehabes commented on js/deer-config.js:
```diff
@@ -421,7 +424,7 @@ export default {
                         li += `>
                             ${inclusionBtn}
-                            <a target="_blank" href="${options.link}${glossID}">
+                            <a target="_blank" href="${deduplicatedList.link}${glossID}">
```
This is causing bad links with undefined on gloss-transcription.html.  I think you want `options.link` here.
`
```HTML
<a target="_blank" href="undefinedhttps://devstore.rerum.io/v1/id/61436336e74876243131a4ea">
    <span class="serifText">Docens suos ascendere</span>
</a>
```

thehabes reviewed (approved):
Way to stick with it.  Relax on your time off @TheTuTran 

🎉🎉🎉 ─=≡Σ((( つ◕ل͜◕)つ

> ⊹╰(⌣ʟ⌣)╯⊹

AbhilashKotha reviewed (approved):
lgtm
